src/main/java/de/tum/in/www1/artemis/web/rest/PersonalAccessTokenResource.java

```diff
@@ -0,0 +1,79 @@
+package de.tum.in.www1.artemis.web.rest;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Profile;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.web.bind.annotation.*;
+
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.repository.UserRepository;
+import de.tum.in.www1.artemis.security.UserNotActivatedException;
+import de.tum.in.www1.artemis.security.jwt.TokenProvider;
+import de.tum.in.www1.artemis.web.rest.util.HeaderUtil;
+
+/**
+ * Controller to request personal access tokens.
+ */
+@RestController
+@RequestMapping("/api")
+@Profile("pat")
+public class PersonalAccessTokenResource {
+
+    private static final String ENTITY_NAME = "personalAccessTokenResource";
+
+    @Value("${jhipster.clientApp.name}")
+    private String applicationName;
+
+    private final TokenProvider tokenProvider;
+
+    private final UserRepository userRepository;
+
+    @Value("${artemis.personal-access-token.max-lifetime-in-days:30}")
+    private long personalAccessTokenMaxLifetimeDays;
+
+    public PersonalAccessTokenResource(TokenProvider tokenProvider, UserRepository userRepository) {
+        this.tokenProvider = tokenProvider;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Generates a personal access token (JWT token) for the current user.
+     *
+     * @param lifetimeDays the lifetime for the JWT token in days
+     * @return a JWT Token
+     */
+    @PreAuthorize("hasRole('USER')")
```

Comment:
In terms of security, I only have only concern:
With the token, the user gets full access.

So technically, when someone gets access to the token, they could renew it whenever they want and keep permanent access. Unless that person continuously uses PAT, they will never recognise that their account has been highjacked.

So, in my opinion, PAT should be significantly different from the token generated on login so that we can secure other functionality.

Also, there is no reason why a PAT should allow to change personal settings, passwords, etc.

In my opinion, we should limit the access of a PAT to only what is really necessary and in the future we might even be able to support certain settings if we really need different powerful access tokens.

Reply:
Thanks for the comments!

You are right that tokens cannot be invalidated and therefore can be renewed until the the user is deactivated or the secret is changed (to invalidate all tokens)

The password can't be changed (Unless you are an admin) since the endpoint needs the old password.

> In my opinion, we should limit the access of a PAT to only what is really necessary and in the future we might even be able to support certain settings if we really need different powerful access tokens.

Depends on the use cases. We would use this for programming tasks and grading.

Reply:
Good to know regarding the password, however, that's the less dangerous thing about it anyway, as the user would immediately know that the account is compromised. The real danger here lies with the renewal.

> Depends on the use cases. We would use this for programming tasks and grading.

Yes, in that case, you can enable all related endpoints for PATs and not allow everything else. And if someone wants to extend this in the future, they can do so, but in the meantime, we limit what an attacker could do with a compromised account.